File: CAS/benchmark.rb
#require 'stringio'
require 'SymDesc'			# cas
require_relative 'ODE'	
require_relative 'test'
require 'ruby-prof'	# profiling
require 'csv'
require 'memory_profiler'

x, y = var :x, :y
function_1 = 48*x**3 - 84*x**2 + 42*x - 36
function_2 = -4*x**3  + 44*x - 10*x**2 - 30
function = -2*x*y**2
x_0 = 0
y_0 = 1
x_end = 1
order = 4
step_size = 0.001
#p primitiveEuclidean(function_1,function_2)
# profile the code
# !!!!!!!!!!!!!!!!!!!!!!!!!!1 = change name of the file 
CSV.open("rb.csv","w+") do |csv|
	csv << ["n_calls,total_tt,mem_alloc,objs"]
	for i in (0..4999) do
		#p i
		# ... code to profile ...
		RubyProf.start
		taylor(function, order, step_size)
		#ODE_eval(function, x_0, y_0, x_end, order, step_size)
		#primitiveEuclidean(function_1,function_2)
		result = RubyProf.stop

		MemoryProfiler.start
		taylor(function, order, step_size)
		#ODE_eval(function, x_0, y_0, x_end, order, step_size)
		#primitiveEuclidean(function_1,function_2)
		report = MemoryProfiler.stop
		objects = report.total_allocated #numero di oggetti
		mem_alloc = report.total_allocated_memsize #bytes

		calls = result.threads[0].methods
		n_calls = 0
		total_tt = 0.0

		calls.each do |numberofcalls|
			n_calls = n_calls + numberofcalls.called
			total_tt = total_tt + numberofcalls.self_time()
		end
		csv << [n_calls, total_tt, mem_alloc, objects]
	end
end

File: CAS/benchmark.py
import cProfile # profiling
import pstats 	# statistical display
import io				#	input - output
import sympy as sym
import csv
from ODE import taylor, ODE_eval, x, y
from sympy import gcdex
from sympy.abc import x
import numpy as np
from guppy import hpy
import psutil


# variables used
function_1 = 48*x**3 - 84*x**2 + 42*x - 36
function_2 = -4*x**3  + 44*x - 10*x**2 - 30
function = -2*x*y**2
x_0 = 0
y_0 = 1
x_end = 1
order = 4
step_size = 0.001

# open csv file
with open('py.csv', 'w+') as file:
	# write headers on file
	file.write("n_calls,total_tt,mem_alloc,objs\n")
	# insert expressione / function here	
	for i in np.arange(0,5000):
		print(i)   # uncomment to follow the index of the loop
		# for memory allocation-------------------------------------------
		h = hpy()
		h.setrelheap()
		# uncomment one of the following
		#taylor(function, order, step_size)
		ODE_eval(function, x_0, y_0, x_end, order, step_size)
		#gcdex(function_1, function_2)
		f = h.heap()			# report saved as a var
		# end of memory allocation----------------------------------------
		# memory allocation
		memory_alloc = f.size
		# objects allocated
		objects = str(f).split('\n')[0].split(' ')[5]

		# for timing------------------------------------------------------
		pr = cProfile.Profile()
		pr.enable()
		# uncomment one of the following
		#taylor(function, order, step_size)
		ODE_eval(function, x_0, y_0, x_end, order, step_size)
		#gcdex(function_1, function_2)
		pr.disable()
		#end of profiing--------------------------------------------------

		# manage profiler report to gather timing
		s = io.StringIO()
		ps = pstats.Stats(pr, stream=s)#.sort_stats(sortby)
		ps.print_stats()
		# write on file
		file.write(str(ps.total_calls) + "," + str(ps.total_tt) + "," + str(memory_alloc) + "," + objects + "\n")
